Reference documentation for add_data_node and delete_data_node (#171)
timescale/docs.timescale.com-content: api.md

SELECT add_dimension('conditions', 'device_id', number_partitions => 2, if_not_exists => true);
```

---

> ## add_data_node() [](add_data_node)

Reviewer: For this type of distributed hypertable / distributed TimescaleDB management, I think we need to create some canonical instructions about where they are executing, and possibly even capture this with the arguments.  For example, I *think* this is executed on the access node which is handling DDL for the distributed hypertable, but this should be clarified?

Also, I find it gets a little confusing between "database" and "bootstrap_database".  Both seem to refer to the name on the remote data node where the the distributed hypertable will be instantiated.

Similarly, to nit pick, are we _creating_ a "remote" hypertable, or is this about expanding an existing distributed hypertable?

Finally, I don't fully understand IP address / hostname management from this function, given that hostname appears to be _optional_ from these instructions.  What happens if the hostname isn't provided?  Minor q:  Is there support for submitting IP addresses vs. hostnames (or is it just passing in IPv4/IPv6 as a string to host argument?)

Reviewer: Now reading the same usage below, seems like this is meant to be used _below_ we actually create the distributed hypertable.  I think this could be better clarified in the preface to this API call?

Similarly, please clarify is this is _also_ used for elastically expanding an existing hypertable (i.e., you can in the middle of existing hypertable, add_data_node() then expand hypertable to that data node), or if that's a different API call.

Author: This is executed on the access node and yes, there are no general instructions on how distributed hypertables work, which would help in making the role of this particular function clear.

I agree that all the users, passwords, and hosts is a little confusing. This seems to be an intermediate situation and will be resolved as we go forward. IMHO, we should still document it in it's current state though.

The function `attach_data_node()` is used to elastically scale an existing hypertable, as clarified by @erimatnor below. Followed his suggestion and updated the text.

Reviewer 2: @mkindahl Note PR here: https://github.com/timescale/timescaledb-private/pull/193

It refactors and simplifies this function. Also, it generally hardens the server management functions with proper permissions checks. 

You need USAGE on servers to add servers and owner on a hypertable to be able to attach servers to it. Might also be worth mentioning.

Author: @erimatnor Yeah, but I think we can merge this PR in the current state and fix things as we go, adding a documentation task after the PR above is merged.

> ## add_data_node() :community_function: [](add_data_node)

Reviewer: I do not believe licensing has been finalized.  Would like to pause merging (although can fix below) until finalized, or possibly just remove the decorator until we do (will get back to you today about what to do here).  Sorry for mixed messages. :)

Author: Ok. Sitting on the commit for the time being then.

Author: From what I understand, the decision was that the decorator should not be there, so removing it.

> WHen executed on the access node, this will add a new data node to the

Reviewer 3: Am I correct that adding the first data node makes the current node to become an access node? Can it be clarified what it means to call it on the node the first time?

Reviewer 3: This might be good place to give idea to users what an access node is, or at a function description where it is created.

Reviewer 3: Double capital in `WHen`

Author: @k-rus There is a need for a separate description of what access nodes and data nodes are, but I think this is beyond the scope of this (reference) documentation.

IMHO, nodes as such as not dedicated data or access nodes, so it does not "become" an access node. For a distributed hypertable, there is one access node where the table is accessed as a whole, and data nodes where the pieces of the table is, so a node can be both an access node and a data node depending on how the hypertable was created.

Reviewer 3: I have the same understanding. Thus the phrase

> When executed on the access node

is confusing for me. Can it be better something like:

> Add a data node into the current database to be used with distributed hypertables through this access node.

Author: Reformulated.

use [`attach_data_node()`](#attach_data_node).

The function will attempt to bootstrap the data node by creating the
database where the remote hypertables will later be created and also
load the TimescaleDB extension on the data node.

#### Errors

* An error will be given if an attempt is made to run the function
  inside a transaction.
* An error will be given if the remote database already exists on the
  data node and `if_not_exists` is not set.

#### Required Arguments [](add_data_node-required-arguments)

| Name        | Description            |
| ----------- | -----------            |
| `node_name` | Name of the data node. |

##### Parameter Description

* `node_name` - Name to use for the node locally.

#### Optional Arguments [](add_data_node-optional-arguments)

| Name                 | Description                                           |
|----------------------|-------------------------------------------------------|
| `host`               | Host name for the remote data node.                   |
| `port`               | Port to use on the remote data node.                  |
| `database`           | Database name where the remote hypertable is created. |

> | `local_user`         | User name on the local node.                          |

Reviewer 3: Should it be:
> User name on the local access node.
I.e., similar to the description of `remote_user`?

| `remote_user`        | User name on the remote data node.                    |
| `password`           | Password on the remote data node.                     |

> | `if_not_exists`      | Do not fail if the hyper table already exists         |

Reviewer 3: As it described above the functionality doesn't affect existing hyper tables. Is it about existing database? Or do I miss anything?

Author: Fixed.

Reviewer 3: @mkindahl @erimatnor Can you explain which hypertable is meant here?

| `bootstrap_database` | Database to bootstrap.                                |
| `bootstrap_user`     | User to used for bootstrapping.                       |

> | `bootstrap_password` | Password for user use when bootstrapping.             |

Author: Fixed.

##### Parameter Description

* `host` - the name of the host to use for the remote data node. This

>   defaults to `'localhost'`.

Reviewer 3: Is it possible to use IP instead of the host name?

Author: Since `getaddrinfo(3)` is used, I would say this is implicit?

Reviewer 3: It looks like it is common for Postgres documentation to say just host name, meaning both name and IP. So I guess it is obvious for the users.

> * `post` - Port to use for the remote data node. This defaults to the

Reviewer 3: @mfreed Good point. I feel the call with defaults should always fail. The default host is the local host and the default database is the same database name as current, thus it will try to add the same node/database to itself, which will fail if `if_not_exists` is not set.

> * `post` - Port to use for the remote data node. This defaults to the

Reviewer: I guess I should have been clearer:  From this discussion, my reaction is that `host` should be *required* not optional.

> * `post` - Port to use for the remote data node. This defaults to the

Author: I think it will be confusing, yes. Since data nodes are really only sensible if elsewhere, we can have the host and port as required arguments. If a user really want to have things locally, she can always provide the local address herself.

> * `post` - Port to use for the remote data node. This defaults to the

Reviewer: I think with port as optional makes sense given default to postgres, was more commenting about host field.

> * `post` - Port to use for the remote data node. This defaults to the

Author: Right.

> * `post` - Port to use for the remote data node. This defaults to the

Author: Will handle this as a separate PR.

  default PostgreSQL port.

* `database` - Name of the database to use on the remote data
  node. Default is to use the current database in use when calling the
  function.

* `local_user` - Name of the local user. Default is to use the current
  user.

* `remote_user` - Name of the user on the remote data node. Default is
  to use the same user as `local_user`.

> * `password` - Password to use on the remote data node. Defaults is to

Author: Fixed.

  not use a password.

> 

Reviewer 3: Is the description of `if_not_exists` missing.

Author: Added.

* `bootstrap_database` - The database to use when bootstrapping the

>   server as described above. The default is to use `'postgres'`.

Reviewer 3: Is it possible to provide hyperlink to the description above?

Author: Does not seem to be easy to add labels arbitrarily.

* `bootstrap_user` - User to use when bootstrapping the server as

>   described above. The default is to use the saver provided as

Reviewer: To be honest, I don't think I clearly understand from this writeup the difference between the "bootstrap_user" and the "remote_user"?

>   described above. The default is to use the saver provided as

Author: `remote_user` was removed in the latest iteration. Will update documentation.

  `remote_user` (or the default).

* `bootstrap_password` - Password to use when bootstrapping the server
  as described above. The default is to use the password as provided
  as `password`.

#### Returns [](add_data_node-returns)

| Column          | Description                                   |
|-----------------|-----------------------------------------------|

> | `node_name`     | Name of the data node                         |

Reviewer: remote data node?

Author: Well... hmm... here are some confusions. The name is used locally to refer to a remote data node. On another access node, the same data node might have a different name. Changing to "local name to use for data node".

| `username`      | Local username                                |
| `node_username` | Remote username                               |

> | `created`       |                                               |

Reviewer: What is type of created?  It this a boolean or time?

Reviewer 2: `created` is  a boolean that indicates whether the remote database was created or not (needed because of `if_not_exists`parameter).

Author: Fixed.

| `username`      | Local username                                |
| `node_username` | Remote username                               |

> | `created`       | True if a new table was created               |

Reviewer 3: There is nothing above in the description of the function indicates that a new table will be created. Isn't it a database created?

Author: Yes, my fault. Fixed.

#### Sample usage [](add_data_node-examples)

> Let's assume that you have an existing table `conditions` and want to

Reviewer: Assume that this should be "hypertable" `conditions"?   And below "chunks of the hypertable"?

Author: Changed this.

use the `time` column as the time partitioning column and the

> `location` column as the space partitioning column. You also want want

Reviewer 3: Duplication of `want`

Author: Fixed.

SELECT add_dimension('conditions', 'device_id', number_partitions => 2, if_not_exists => true);
```

---
## add_data_node() [](add_data_node)

> Add a new data node to the current database for use with distributed

Reviewer 4: Could we add a sentence here that says something along the lines of - this command must be run on an access node?

Author: Done.

hypertables.

This includes bootstrapping the data node by creating and configuring

> a new database on it. Newly created distributed hypertables will be

Author: Since bootstrapping is done, I think we need to describe it somewhere. The text here, however, might be unnecessarily terse.

> a new database on it. Newly created distributed hypertables will be

Author: I added a brief description about this below, so resolving this comment. Please tell me if you think I should change the text.

able to make use of this data node for distributing data across the
cluster.

> Note, however, that already existing distributed hypertables may

Author: Yeah, I have a tendency to add "however" too often. =)

Reformulated the sentence as well. PTAL.

require a separate `attach_data_node()` call to schedule the data node
for use with those tables as they do not automatically expand to
include the new data node.

The function will attempt to bootstrap the data node by creating the
database where the remote hypertables will later be created and also

> load the TimescaleDB extension on the data node.

Reviewer 4: Ah, I like this explanation of bootstrapping here, so definitely think we can remove the bootstrapping reference above

Author: Done.

SELECT add_dimension('conditions', 'device_id', number_partitions => 2, if_not_exists => true);
```

---
## add_data_node() [](add_data_node)

Add a new data node to the database to be used for creating
distributed hypertables.

Newly created distributed hypertables will be able to make use of this
data node for distributing data across the cluster. Note that existing
distributed hypertables will not use the newly added data node. For
existing distributed hypertables to use added data nodes, you need to
use [`attach_data_node()`](#attach_data_node).

The function will attempt to bootstrap the data node by creating the
database where the remote hypertables will later be created and also
load the TimescaleDB extension on the data node.

#### Errors

* An error will be given if an attempt is made to run the function
  inside a transaction.
* An error will be given if the remote database already exists on the
  data node and `if_not_exists` is not set.

#### Required Arguments [](add_data_node-required-arguments)

| Name        | Description                         |
| ----------- | -----------                         |
| `node_name` | Name of the data node.              |

> ##### Parameter Description

Reviewer 5: This section is not neccesary

* `node_name` - Name to use for the node locally.

#### Optional Arguments [](add_data_node-optional-arguments)

| Name                 | Description                                           |
|----------------------|-------------------------------------------------------|
| `host`               | Host name for the remote data node.                   |
| `port`               | Port to use on the remote data node.                  |
| `database`           | Database name where the remote hypertable is created. |
| `password`           | Password on the remote data node.                     |
| `if_not_exists`      | Do not fail if the data node already exists.          |
| `bootstrap_database` | Database to use when bootstrapping.                   |
| `bootstrap_user`     | User to be used for bootstrapping.                    |
| `bootstrap_password` | Password to be used when bootstrapping.               |

> ##### Parameter Description

Reviewer 5: These descriptions can be incorporated into the table

* `host` - the name of the host to use for the remote data node. The
  default is `'localhost'`.

* `port` - Port to use for the remote data node. The default is the
  default PostgreSQL port.

* `database` - Name of the database to use on the remote data
  node. The default is the current database.

* `password` - Password to be used on the remote data node. The
  default is to not use a password.

* `bootstrap_database` - The database to use when bootstrapping the
  data node as described above.
  
  The default is `'postgres'`.

* `bootstrap_user` - User to be used when bootstrapping the data node

>   as described above. The user need to have `CREATEDB` permissions on

Reviewer 5: ```suggestion
  as described above. The user needs to have `CREATEDB` permissions on
```

Reviewer 6: fixed

use [`attach_data_node()`](#attach_data_node).

The function will attempt to bootstrap the data node by creating the
database where the remote hypertables will later be created and also
load the TimescaleDB extension on the data node.

#### Errors

* An error will be given if an attempt is made to run the function
  inside a transaction.
* An error will be given if the remote database already exists on the
  data node and `if_not_exists` is not set.

#### Required Arguments [](add_data_node-required-arguments)

| Name        | Description            |
| ----------- | -----------            |
| `node_name` | Name of the data node. |

##### Parameter Description

* `node_name` - Name to use for the node locally.

#### Optional Arguments [](add_data_node-optional-arguments)

| Name                 | Description                                           |
|----------------------|-------------------------------------------------------|
| `host`               | Host name for the remote data node.                   |
| `port`               | Port to use on the remote data node.                  |
| `database`           | Database name where the remote hypertable is created. |

> | `local_user`         | User name on the local node.                          |

Reviewer 2: Note the now updated interface that simplifies this interface (removes some args)

Author: Ok. Rechecking `ddl_api.sql`

| `remote_user`        | User name on the remote data node.                    |
| `password`           | Password on the remote data node.                     |
| `if_not_exists`      | Do not fail if the data node already exists.          |

> | `bootstrap_database` | Database to bootstrap.                                |

Reviewer 2: This is not the database to bootstrap but the database to connect to in order to bootstrap (create) another database (given by the `database` argument.)

Author: That is *really* confusing... :-/

> | `bootstrap_user`     | User to be used for bootstrapping.                    |

Reviewer 2: This user must have CREATEDB permissions on the data node and AFAIK be a  superuser to create the extension in the new db

| `bootstrap_password` | Password to be used when bootstrapping.               |

##### Parameter Description

* `host` - the name of the host to use for the remote data node. This
  defaults to `'localhost'`.

> * `port` - Port to use for the remote data node. This defaults to the

Reviewer 2: I think it actually defaults to the port used by the access node, which is usually the default postgres port

Author: Not sure what you want changed here.

Reviewer: I'm thinking he's suggesting that rather than defauling to default PostgreSQL port, it's:  ```* `port` - Port to use for the remote data node. This defaults to the
  port used by the access node```
or 
```by the access node on which the function is executed```

Reviewer 6: Went with "The default is the  PostgreSQL port used by the access node on which the function is  executed."

  default PostgreSQL port.

> * `database` - Name of the database to use on the remote data

Reviewer 2: current database name...

Author: Not sure what change you want here.

Reviewer 6: I assume you were referring to the default value.  It's been fixed.

  node. Default is to use the current database in use when calling the
  function.

> * `local_user` - Name of the local user. Default is to use the current

Reviewer 2: This parameter is no longer present

Author: Removed.

  user.

> * `remote_user` - Name of the user on the remote data node. Default is

Reviewer 2: This parameter is no longer present

Author: Removed.

  to use the same user as `local_user`.

* `password` - Password to be used on the remote data node. Default is
  to not use a password.

* `bootstrap_database` - The database to use when bootstrapping the
  data node as described above. The default is to use `'postgres'`.

* `bootstrap_user` - User to be used when bootstrapping the data node
  as described above. The default is to use the user provided as
  `remote_user` (or the default).

* `bootstrap_password` - Password to be used when bootstrapping the
  data node as described above. The default is to use the password as
  provided as `password`.

* `if_not_exists` - Do not fail the operation if the data node already
  exists. The data node will not be overwritten, but the command will
  not generate an error and instead generate a notice.

#### Returns [](add_data_node-returns)

> 

Reviewer 2: These return columns have been changed

Author: Fixed.

SELECT add_dimension('conditions', 'device_id', number_partitions => 2, if_not_exists => true);
```

---
## add_data_node() [](add_data_node)

Add a new data node to the database to be used for creating
distributed hypertables.

Newly created distributed hypertables will be able to make use of this

> data node for distributing data across the cluster. Note that existing

Reviewer 3: ```suggestion
data node for distributing data across the distributed database. Note that existing
```
As mentioned in another PR it is better to avoid `cluster` due to confusion with PostgreSQL terminology.

Reviewer 6: Changed this to "apportioning data across the distributed database" to make it less repetitive.

1. Connecting to a specified postgres instance
2. Creating the database that will serve as the new data node
3. Loading the TimescaleDB extension on the new database

#### Errors

An error will be given if:
* an attempt is made to run the function inside a transaction _or_
* the remote database already exists on the data node and
  `if_not_exists` is not set.

#### Required Arguments [](add_data_node-required-arguments)

| Name        | Description                         |
| ----------- | -----------                         |
| `node_name` | Name for the data node.             |

#### Optional Arguments [](add_data_node-optional-arguments)

| Name                 | Description                                           |
|----------------------|-------------------------------------------------------|
| `host`               | Host name for the remote data node. The default is `'localhost'`. |
| `port`               | Port to use on the remote data node. The default is the PostgreSQL port used by the access node on which the function is executed. |
| `database`           | Database name where remote hypertables will be created. The default is the current database name. |

> | `password`           | Password to be used on the remote data node. The default is to not use a password. |

Reviewer 3: Will it create new user in the remote database with the name of the current local user and the given remote password? Is it good idea to clarify this? Or is it obvious for PostgreSQL users?

Reviewer 2: Ideally, it would create the user, but it currently doesn't because we haven't implemented any of the roles/permissions management. So, for now I think we can just add a note that the user is assumed to exist on data nodes and that it is a manual process right now.

Reviewer 6: Added a note here "Password to be used when connecting to the remote data node. Note that the user name will be the same as the current user. The default is to not use a password."  There probably is still some room for confusion here, but hopefully this is better.

| `bootstrap_password` | Password to be used when bootstrapping. The default is the password provided in `password`. |

#### Returns [](add_data_node-returns)

| Column              | Description                                       |
|---------------------|---------------------------------------------------|
| `node_name`         | Local name to use for the data node               |
| `host`              | Host name for the remote data node                |
| `port`              | Port for the remote data node                     |
| `database`          | Name of database used on the remote data node     |
| `node_created`      | Was the node created                              |
| `database_created`  | Was the database created on the remote data node  |
| `extension_created` | Was the extension created on the remote data node |

#### Sample usage [](add_data_node-examples)

Let's assume that you have an existing hypertable `conditions` and
want to use `time` as the time partitioning column and `location` as
the space partitioning column. You also want to distribute the chunks
of the hypertable on two data nodes `dn1.example.com` and

> `dn2.example.com`:

Reviewer 3: Isn't it good to mention that the current user should be a superuser to succeed? Otherwise I am confused about superuser requirement for bootstrap user.

Reviewer 6: Yeah, I added a note and provided a more fleshed out example.

SELECT add_dimension('conditions', 'device_id', number_partitions => 2, if_not_exists => true);
```

---
## add_data_node() [](add_data_node)

Add a new data node to the database to be used for creating
distributed hypertables.

Newly created distributed hypertables will be able to make use of this
data node for distributing data across the cluster. Note that existing
distributed hypertables will not automatically use the newly added
data node. For existing distributed hypertables to use added data

> nodes, use [`attach_data_node()`](#attach_data_node).

Reviewer 5: ```suggestion
nodes, use [`attach_data_node`](#attach_data_node).
```

The function will attempt to bootstrap the data node by:
1. Connecting to a specified postgres instance
2. Creating the database that will serve as the new data node
3. Loading the TimescaleDB extension on the new database

#### Errors

An error will be given if:

> * an attempt is made to run the function inside a transaction.

Reviewer 5: ```suggestion
* an attempt is made to run the function inside a transaction _or_
```

| `bootstrap_password` | Password to be used when bootstrapping. The default is the password provided in `password`. |

#### Returns [](add_data_node-returns)

| Column              | Description                                       |
|---------------------|---------------------------------------------------|
| `node_name`         | Local name to use for the data node               |
| `host`              | Host name for the remote data node                |
| `port`              | Port for the remote data node                     |
| `database`          | Name of database used on the remote data node     |
| `node_created`      | Was the node created                              |
| `database_created`  | Was the database created on the remote data node  |
| `extension_created` | Was the extension created on the remote data node |

#### Sample usage [](add_data_node-examples)

Let's assume that you have an existing hypertable `conditions` and

> want to use the `time` column as the time partitioning column and the

Reviewer 5: ```suggestion
want to use `time` as the time partitioning column and
```

> `location` column as the space partitioning column. You also want to

Reviewer 5: ```suggestion
`location` as the space partitioning column. You also want to
```

Reviewer 6: done

distribute the chunks of the hypertable on two data nodes

> `dn1.example.com` and `dn2.example.com`.

Reviewer 5: ```suggestion
`dn1.example.com` and `dn2.example.com`:
```

> To do this, you can just type:

Reviewer 5: ```suggestion

```

| `if_not_exists`      | Do not fail if the data node already exists. The data node will not be overwritten, but the command will not generate an error and instead generate a notice. |
| `bootstrap_database` | Database to use when bootstrapping as described above. The default is `'postgres'`. |
| `bootstrap_user`     | User to be used for bootstrapping. This user needs to have `SUPERUSER` permissions on the remote database server, unless the remote database already exists with the timescaledb extension. The default is the current user. |
| `bootstrap_password` | Password to be used when bootstrapping. The default is the password provided in `password`. |

#### Returns [](add_data_node-returns)

| Column              | Description                                       |
|---------------------|---------------------------------------------------|
| `node_name`         | Local name to use for the data node               |
| `host`              | Host name for the remote data node                |
| `port`              | Port for the remote data node                     |

> | `database`          | Name of database used on the remote data node     |

Reviewer 3: ```suggestion
| `database`          | Database name used on the remote data node     |
```
Will look more consistent with the above. Otherwise it is good to use article. But I am not so good in English.

Reviewer 6: changed

> | `node_created`      | Was the node created                              |

Reviewer 2: ```suggestion
| `node_created`      | Was the data node created locally                              |
```

Reviewer 6: changed

| `database_created`  | Was the database created on the remote data node  |
| `extension_created` | Was the extension created on the remote data node |

#### Sample usage [](add_data_node-examples)

Let's assume that you have an existing hypertable `conditions` and
want to use `time` as the time partitioning column and `location` as
the space partitioning column. You also want to distribute the chunks
of the hypertable on two data nodes `dn1.example.com` and
`dn2.example.com`:

```sql
SELECT add_data_node('dn1', host => 'dn1.example.com');
SELECT add_data_node('dn2', host => 'dn2.example.com');
SELECT create_distributed_hypertable('conditions', 'time', 'location');
```

Note that the previous example will only succeed if the current user
has `SUPERUSER` permission on the two remote systems and doesn't
require a password to connect.  If the current user doesn't have the
appropriate permissions, you'll have to provide a bootstrap user as
described above:

```sql

> SELECT add_data_node('dn1', host => 'dn1.example.com', password => 'mydn1password', bootstrap_user => 'dn1_superuser', bootstrap_password => 'dn1_su_password');

Reviewer 3: May be split into several lines?

```

---
### add_data_node() [](add_data_node)

> Create a new data node for use when creating distributed hypertables.

Reviewer 2: I would avoid the "create" terminology here since people might get the wrong idea of what this function does. I would say something like:

"Add a new data node to the current database for use with distributed hypertables. This includes bootstrapping the data node by creating and configuring a new database on it. Newly created distributed hypertables will be able to make use of this data node for distributing data across the cluster. Note, however, that already existing distributed hypertables may require a separate `attach_data_node()` call to schedule the data node for use with those tables as they do not automatically expand to include the new data node."

To create a distributed hypertable it is necessary to define the data

> nodes where the data is going to be stored, which this utility is used

Reviewer 2: utility -> function

for.

> The utility will attempt to bootstrap the data node by creating the

Reviewer 2: I would elide this part if you go with my suggestion above.

Author: Yup.

database where the remote hypertables will later be created and also
load the TimescaleDB extension on the data node.

#### Errors

* An error will be given if an attempt is made to run the function
  inside a transaction.

> * An error will be given if the database already exists on the data

Reviewer 2: remote database

Author: Fixed.

  node and `if_not_exists` is not set.

#### Required Arguments [](add_data_node-required-arguments)

| Name        | Description       |
| ----------- | -----------       |

> | `node_name` | Name of the node. |

Reviewer 2: Name of the data node.

Author: Done.

##### Parameter Description

* `node_name` - Name to use for the node locally.

#### Optional Arguments [](add_data_node-optional-arguments)

| Name                 | Description                                           |
|----------------------|-------------------------------------------------------|

> | `host`               | Host name for the remote data node.                   |

Reviewer 2: Should we put the default setting in this table instead of below? Or, is this the established pattern?

Author: The tables become exceedingly wide and the default is occasionally derived from other parameters, hence require more elaborate descriptions, which is why I went with this approach.

| Column          | Description                                   |
|-----------------|-----------------------------------------------|
| `node_name`     | Name of the data node                         |
| `host`          | Host name for the remote data node            |
| `port`          | Port for the remote data node                 |
| `database`      | Name of database used on the remote data node |
| `username`      | Local username                                |
| `node_username` | Remote username                               |
| `created`       |                                               |

#### Sample usage [](add_data_node-examples)

Let's assume that you have an existing table `conditions` and want to
use the `time` column as the time partitioning column and the
`location` column as the space partitioning column. You also want want
to distribute the chunks of the table on two data nodes
`dn1.example.com` and `dn2.example.com`.

To do this, you can just type:

```sql
SELECT add_data_node('dn1', host => 'dn1.example.com');
SELECT add_data_node('dn2', host => 'dn2.example.com');
SELECT create_distributed_hypertable(
    'conditions', 'time', 'location',

> 	servers => '{"dn1","dn2"}'

Reviewer 2: You may elide the server list here and it will make use of all servers in the database. I think this should probably be the default documented behavior. We can go into more details about servers  (should probably be `data_nodes`) in the description of `create_distributed_hypertable()`.

Author: Done.

  WITH (timescaledb.transaction_per_chunk);
```

## delete_data_node() [](delete_data_node)

This function will remove the data node locally. This will *not*
affect the remote database in any way, it will just update the local
index over all existing data nodes.

The data node will be detached from all hypertables that are using

> it. For more information, see [`detach_data_node`](#detach_data_node).

Reviewer 2: ```suggestion
it if permissions and data integrity requirements are satisfied. For more information, see [`detach_data_node`](#detach_data_node).
```

Reviewer 6: changed.

  WITH (timescaledb.transaction_per_chunk);
```

## delete_data_node() [](delete_data_node)

This function will remove the data node locally. This will *not*
affect the remote database in any way, it will just update the local
index over all existing data nodes.

The data node will be detached from all hypertables that are using
it if permissions and data integrity requirements are satisfied. For
more information, see [`detach_data_node`](#detach_data_node).

#### Errors

An error will be generated if the data node cannot be detached from
all attached hypertables.

#### Required Arguments [](delete_data_node-required-arguments)

| Name        | Description            |
| ----------- | -----------            |
| `node_name` | Name of the data node. |

#### Optional Arguments [](delete_data_node-optional-arguments)

| Name        | Description                                           |
|-------------|-------------------------------------------------------|
| `if_exists` | Prevent error if the data node does not exist. Defaults to false. |

> | `cascade`   | Cascade the delete so that all objects dependent on the data node are deleted as well. Defaults to false. |

Reviewer 3: I have no idea what the dependent objects. Will it be important for users to know?

  WITH (timescaledb.transaction_per_chunk);
```

## delete_data_node() [](delete_data_node)

This function will remove the data node locally. This will *not*
affect the remote database in any way, it will just update the local
index over all existing data nodes.

The data node will be detached from all hypertables that are using

> it. For more information, see [`detach_data_node`].

Reviewer 5: ```suggestion
it. For more information, see [`detach_data_node`](#detach_data_node).
```

#### Errors

> * An error will be generated if the data node cannot be detached from

Reviewer 5: ```suggestion
An error will be generated if the data node cannot be detached from
```

>   some hypertable.

Reviewer 5: ```suggestion
all attached hypertables.
```

#### Required Arguments [](delete_data_node-required-arguments)

| Name        | Description            |
| ----------- | -----------            |
| `node_name` | Name of the data node. |

##### Parameter Description

* `node_name` - Name to use for the node locally.

#### Optional Arguments [](delete_data_node-optional-arguments)

| Name        | Description                                           |
|-------------|-------------------------------------------------------|

> | `if_exists` | Don't generate error if the data node does not exist. |

Reviewer 5: ```suggestion
| `if_exists` | Prevent error if the data node does not exist. |
```

The data node will be detached from all hypertables that are using
it. For more information, see [`detach_data_node`].

#### Errors

* An error will be generated if the data node cannot be detached from
  some hypertable.

#### Required Arguments [](delete_data_node-required-arguments)

| Name        | Description            |
| ----------- | -----------            |
| `node_name` | Name of the data node. |

##### Parameter Description

* `node_name` - Name to use for the node locally.

#### Optional Arguments [](delete_data_node-optional-arguments)

| Name        | Description                                           |
|-------------|-------------------------------------------------------|
| `if_exists` | Don't generate error if the data node does not exist. |
| `cascade`   | Cascade the delete.                                   |
| `force`     | Force removal of data nodes from hypertables.         |

##### Parameter Description

* `if_exists` - Do not generate an error if the data node does not
  exist. Defaults to false.

* `cascade` - Cascade the delete so that all objects dependent on the
  data node are deleted as well. Defaults to false.

> * `force` - Force removal of data nodes from hypertables, potentially

Reviewer 2: Data loss should not be possible here. AFAIK, this is only to force removal even if chunks in some hypertables will be under-replicated (this is the same as for detach, since delete means implicit detaches on all hypertables). We never allow removal in case of data loss as that would leave hypertables in an undefined state (basically broken).

Author: Checked the code. You're right, it should only force the operation when there is more than 1 replicated node. Fixed the text.
